### DISCUSSION TITLE
Project Specifications

### DISCUSSION THREAD

**Discussion by TokisanGames** [TokisanGames/Terrain3D#3] [Architecture]
This is a start of goals, but I am going to revamp this list into a a prioritized list of feature specifications.

Priorities correspond to project milestones: 
* Minimum viable product (Mvp)
* Polished product (Pol)
* Nice to have or optional (Opt)

**Overall Goals**
* Fast, quality, stable terrain system
* Easy for level designers to use
* Aligns with goals from [Godot devs](https://github.com/godotengine/godot-proposals/issues/6121) where practical

**Sculpting & Painting**
* [x] Mvp: Operations: Raise, Lower, Flatten, Smooth | Pol: Slope, Erode
* [x] Mvp: Paintable area expands as you paint in other areas
* [x] Mvp: Allows large, broad strokes to quickly form the landscape
* [x] Mvp: Allows fine grained control over small areas
* [x] Mvp: Brush variety, alpha masks 
* [x] Pol: Brush rotates w/ camera
* [x] Opt: Can handle reasonably sharp corners or terraces, rather than just rounded swells.

**Materials**
* [x] Mvp: High number of textures. 16 at least up to 256. See #4.
* [x] Mvp: Color map
* [ ] Pol: Eliminate texture stretching (3D projection)
* [x] Pol: Detiling
* [x] Pol: Auto texturing
* [x] Pol: Dual scaling
* [x] Opt: Paintable UV scale (Witcher 3)
* [ ] Opt: Paintable slope (Witcher 3)

**Mesh**
* [x] Mvp: View distance can stretch to the horizon
* [x] Mvp: Infinite, repeating mesh as the camera moves
* [ ] Mvp: Infinite, repeating heightmap option
* [x] Mvp: Terrain is visually stable without artifacts. Doesn't move, flicker, waver, etc
* [x] Mvp: Adjustable after creation. (Change view distance, lods, limits, scale, mesh resolution, image size)
* [x] Mvp: Supports great heights capable of mountains and valleys
* [x] Opt: Supports holes visually and in collision (if possible)
* [ ] Opt: Tesselation

**Maps**
* [ ] Mvp: Resize maps within the editor
* [x] Mvp: Provides an upgrade path from HTerrain
* [ ] Mvp: Import / export for integration w/ other apps
* [x] Mvp: Basic noise landscape generation
* [ ] Opt: Heightmap generator

**Code/Engine**
* [x] Mvp: Documented, useful API to get information from or manipulate the terrain.
* [x] Pol: Supports navigation mesh baking

**Object Placement**
* [x] Pol: Detail instancing - Painting of no-collision, particles or multimesh instanced meshes or billboards for foliage, small rocks
* [ ] Opt: Scene placement - Painting of PackedScenes which can include collision, scripts, and anything else.

More will be added here. What are your goals or feature specifications you'd like to see?

**Comment by outobugi** [comment]
> Provides an upgrade path from HTerrain

We'll see. Low priority because we need much done before that.

> Can handle reasonably sharp corners or terraces, rather than just rounded swells.

Not hard to do but requires tessellation, or sacrifice vertex budget by having a really dense mesh (~0.3m instead of 1m spacing).

> Terrain is visually stable without artifacts. Doesn't move, flicker, waver, etc

Should be done already if we trust my coding. It was tackled in the [original](https://mikejsavage.co.uk/blog/geometry-clipmaps.html).

> View distance can stretch to the horizon

Almost done. ;)

> Adjustable after creation. (Change view distance, lods, limits, scale, mesh resolution, image size)

Image size could be fixed to 1k. When user paints over that, a new map is generated. (If we go the Juan way)
Need to discuss the distance and how it's implemented. I have some math problems there.


> What are your goals?

To live long enough to see a terrain in Godot.

**Reply by TokisanGames** [reply]
> We'll see. Low priority because we need much done before that.

I cannot use the new system until this is supported. My code already supports it anyway, so it's not difficult.

> Not hard to do but requires tessellation, or sacrifice vertex budget by having a really dense mesh (~0.3m instead of 1m spacing).

TW3 talk a little bit about tessellating the mesh in their video. This could be a nice to have feature. However I do want to be able to support different densities. TW3 does 0.5. We're using 1m, but I want to go down to 0.5 and increase from a 2k to a 4k heightmap.

> Terrain is visually stable without artifacts. Doesn't move, flicker, waver, etc
Yes, hooray. I have this problem now, caused by one mesh rather than individual meshes for each LOD.

>> Adjustable after creation. (Change view distance, lods, limits, scale, mesh resolution, image size)
> Image size could be fixed to 1k. 

By image size I mean the heightmap, colormap, sizes. Mine are not fixed and can be resized at any time through the interface. They can also be independent. The only restriction I have now is power of 2 to align w/ collision.


Next, I will separate out the goals list into more higher and lower priorities. At the top are:
* Mesh generation
* Class structure
* Storage structure
